subsys/mgmt/mcumgr/smp_reassembly.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+#include <net/buf.h>
+#include <mgmt/mcumgr/buf.h>
+#include <mgmt/mcumgr/smp.h>
+#include <mgmt/mgmt.h>
+#include <smp/smp.h>
+
+void zephyr_smp_reassembly_init(struct zephyr_smp_transport *zst)
+{
+	zst->__reassembly.current = NULL;
+	zst->__reassembly.expected = 0;
+}
+
+int zephyr_smp_reassembly_expected(const struct zephyr_smp_transport *zst)
+{
+	if (zst->__reassembly.current == NULL) {
+		return -EINVAL;
+	}
+
+	return zst->__reassembly.expected;
+}
+
+int zephyr_smp_reassembly_collect(struct zephyr_smp_transport *zst, const void *buf, uint16_t len)
+{
+	/* No allocated buffer indicates that we are expecting first fragment */
+	if (zst->__reassembly.current == NULL) {
+		if (len >= sizeof(struct mgmt_hdr)) {
+			uint16_t expected = sys_be16_to_cpu(((struct mgmt_hdr *)buf)->nh_len);
+
+			/*
+			 * Size of header is not included in length found in header but we count
+			 * all expected bytes.
+			 */
+			expected += sizeof(struct mgmt_hdr);
+
+			/* Joining net_bufs not supported yet */
+			if (len > CONFIG_MCUMGR_BUF_SIZE || expected > CONFIG_MCUMGR_BUF_SIZE) {
+				return -ENOSR;
+			}
+
+			if (len > expected) {
+				return -EOVERFLOW;
+			}
+
+			zst->__reassembly.current = mcumgr_buf_alloc();
+			if (zst->__reassembly.current != NULL) {
+				zst->__reassembly.expected = expected;
+			} else {
+				return -ENOMEM;
+			}
+		} else {
+			/* Not enough data to even collect header */
+			return -ENODATA;
+		}
+	}
+
+	/* len is expected to be > 0 */
+	if (zst->__reassembly.expected >= len) {
+		net_buf_add_mem(zst->__reassembly.current, buf, len);
+		zst->__reassembly.expected -= len;
+	} else {
+		/*
+		 * Fragment is longer than expected size of fragments to receive and will not
```

Comment:
Fragment is longer than the expected size and would not fit into the buffer.

Reply:
Fixed